src/cargo/ops/cargo_compile.rs

```diff
@@ -165,33 +165,66 @@ pub fn compile_pkg<'a>(package: &Package,
         (packages, resolved_with_overrides, registry.move_sources())
     };
 
-    let pkgid = match spec {
-        Some(spec) => try!(resolve_with_overrides.query(spec)),
-        None => package.package_id(),
+    let mut invalid_spec = vec![];
+    let pkgids = if spec.len() > 0 {
+        spec.iter().map(|&p| {
+            let id = resolve_with_overrides.query(p);
+            if id.is_err() {
+                invalid_spec.push(p);
+            }
+            id
+        }).filter(|id|
+            id.is_ok()
+        ).map(|id| id.unwrap()).collect::<Vec<&PackageId>>()
+    } else {
+        vec![package.package_id()]
     };
+
+    if spec.len() > 0 && invalid_spec.len() > 0 {
+        return Err(human(format!("could not find package matching spec `{}`",
+                                 invalid_spec.join(", "))));
+    }
+
+    let pkgid = pkgids[0];
     let to_build = packages.iter().find(|p| p.package_id() == pkgid).unwrap();
```

Comment:
Isn't this variable obsolete now? There could be more than one package being built?